mlrun/projects/project.py

```diff
@@ -3446,6 +3448,123 @@ def remove_custom_packager(self, packager: str):
+        :param description (str): description for the API Gateway (optional).
+        :param username (str): Username for authentication (optional).
+        :param password (str): Password for authentication (optional).
+        :param canary (List[int]): List containing canary configuration for each function in the function
+            list respectively (optional).
+            For example, if the function list is [f1, f2], the canary list should have two elements,
+            such as [20, 80]. In this case, 20 represents the percentage of traffic going to f1, and 80 represents
+            the percentage of traffic going to f2.
+
+        @return: API Gateway object if successful, else None.
+        """
+        if not name:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "API Gateway name cannot be empty"
+            )
+        if not isinstance(functions, list):
+            functions = [functions]
+
+        # validating functions
+        if not 1 <= len(functions) <= 2:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                f"Gateway can be created from one or two functions, "
+                f"but the number of functions is passed {len(functions)}"
+            )
+        for func in functions:
+            function_name = (
+                func.metadata.name if hasattr(func, "metadata") else func.name
+            )
+            if func.kind not in mlrun.runtimes.RuntimeKinds.nuclio_runtimes():
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    f"Input function {function_name} is not a Nuclio function"
+                )
+            if func.metadata.project != self.name:
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    f"input function {function_name} "
+                    f"does not belong to this project"
+                )
+
+        # validating canary
+        if canary:
+            if len(functions) != len(canary):
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "Function and canary lists lengths do not match"
+                )
+            for canary_percent in canary:
+                if canary_percent < 0 or canary_percent > 100:
+                    raise mlrun.errors.MLRunInvalidArgumentError(
+                        "The percentage value must be in the range from 0 to 100"
+                    )
+            if sum(canary) != 100:
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "The sum of canary function percents should be equal to 100"
+                )
+        if username and not password:
+            raise mlrun.errors.MLRunInvalidArgumentError("Password is not specified")
+
+        if password and not username:
+            raise mlrun.errors.MLRunInvalidArgumentError("Username is not specified")
+
+        function_names = [func.uri for func in functions]
+        ok = mlrun.db.get_run_db().create_api_gateway(
```

Comment:
If you're changing to create an `APIGateway` here and passing the object to the creation call, I think the validations done above (good coverage, BTW!) should go to the `APIGateway` class - this way, the code here would either get an object which is already validated, or would (if you remain with getting the parts of the GW and constructing it here) create the object here, and would fail if something is wrong. 
It makes more sense since the validation logic belongs to the API GW class, not to the project class.